src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -0,0 +1,403 @@
+                _watcherThread.Join();
+            }
+
+            // Clean up the EventStream, if it exists
+            if (_eventStream != IntPtr.Zero)
+            {
+                Interop.CoreFoundation.CFRelease(_eventStream);
+            }
+
+            // Reset back to zero
+            _eventStream = IntPtr.Zero;
+            _watcherRunLoop = IntPtr.Zero;
+        }
+
+        private void RestartStream()
+        {
+            if (_eventStream != IntPtr.Zero)
+            {
+                // We don't need to rebuild the stream since the path is the same so just restart the stream.
+                if (Interop.CoreServices.FSEventStreamStart(_eventStream) == false)
+                {
+                    // Pass the error back to the user and set the last watched directory to null so that we'll
+                    // cleanup and recreate our streams next time Start is called.
+                    OnError(new ErrorEventArgs(new IOException("", Marshal.GetLastWin32Error())));
+                    _lastWatchedDirectory = String.Empty;
+                }
+            }
+        }
+
+        private void StopStream()
+        {
+            // Just stop the EventStream to be optimistic that we'll be restarted with the same path
+            // and not have to teardown everything and rebuild.
+            if (_eventStream != IntPtr.Zero)
+            {
+                Interop.CoreServices.FSEventStreamStop(_eventStream);
+            }
+        }
+
+        private void WatchForFileSystemEventsThreadStart()
+        {
+            // Get this thread's RunLoop
+            _watcherRunLoop = Interop.CoreFoundation.CFRunLoopGetCurrent();
+            Debug.Assert(_watcherRunLoop != IntPtr.Zero);
+
+            // Schedule the EventStream to run on the thread's RunLoop
+            Interop.CoreServices.FSEventStreamScheduleWithRunLoop(_eventStream, _watcherRunLoop, Interop.CoreFoundation.kCFRunLoopDefaultMode);
+            if (Interop.CoreServices.FSEventStreamStart(_eventStream))
+            {
+                // Notify the StartRaisingEvents call that we are initialized and about to start
+                // so that it can return and avoid a race-condition around multiple threads calling Stop and Start
+                _runLoopStartedEvent.Set();
+
+                // Start the OS X RunLoop (a blocking call) that will pump file system changes into the callback function
+                Interop.CoreFoundation.CFRunLoopRun();
+
+                // When we get here, we've requested to stop so cleanup the EventStream and unschedule from the RunLoop
+                Interop.CoreServices.FSEventStreamStop(_eventStream);
+                Interop.CoreServices.FSEventStreamUnscheduleFromRunLoop(_eventStream, _watcherRunLoop, Interop.CoreFoundation.kCFRunLoopDefaultMode);
+            }
+            else
+            {
+                // An error occurred while trying to start the run loop so fail out
```

Comment:
Good catch, the else block to have a signal there before the OnError event...adding that now